File: src/main/java/com/example/spark/spring/IOCConfiguration.java
package com.example.spark.spring;

import lombok.var;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.streaming.Duration;
import org.apache.spark.streaming.api.java.JavaInputDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import org.apache.spark.streaming.kafka010.ConsumerStrategies;
import org.apache.spark.streaming.kafka010.KafkaUtils;
import org.apache.spark.streaming.kafka010.LocationStrategies;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.util.Collections;
import java.util.HashMap;

@Configuration
public class IOCConfiguration {

    @Bean
    public SparkConf sparkConf() {
        return new SparkConf(true)
                .set("spark.submit.deployMode", "client");
    }

    @Bean
    public JavaSparkContext sc(SparkConf sparkConf) {
        return new JavaSparkContext("local[4]", "spark-kafka-exapmle", sparkConf);
    }

    @Bean
    public SparkSession session(JavaSparkContext sc) {
        return new SparkSession(sc.sc());
    }

    @Bean
    public JavaStreamingContext scc(JavaSparkContext sc) {
        return new JavaStreamingContext(sc, new Duration(2000));
    }

    @Bean
    public JavaInputDStream<ConsumerRecord<String, String>> stream(JavaStreamingContext ssc) {
        var kafkaParams = new HashMap<String, Object>();
        kafkaParams.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "node-3:9092");
        kafkaParams.put(ConsumerConfig.GROUP_ID_CONFIG, "spark.test.group");
        kafkaParams.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
        kafkaParams.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);

        return KafkaUtils.createDirectStream(ssc, LocationStrategies.PreferConsistent(),
                ConsumerStrategies.Subscribe(Collections.singletonList("spark.test"), kafkaParams));
    }
}

File: src/main/scala/com/example/spark/spring/IOCConfiguration.scala
package com.example.spark.spring

import org.apache.kafka.clients.consumer.{ConsumerConfig, ConsumerRecord}
import org.apache.kafka.common.serialization.StringDeserializer
import org.apache.spark.sql.{SQLContext, SparkSession}
import org.apache.spark.streaming.dstream.InputDStream
import org.apache.spark.streaming.kafka010.{ConsumerStrategies, KafkaUtils, LocationStrategies}
import org.apache.spark.streaming.{Duration, StreamingContext}
import org.apache.spark.{SparkConf, SparkContext}
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.{Bean, Configuration}

@Configuration
class IOCConfiguration() {

  @Bean
  def sparkConf(config: SparkProperties): SparkConf = new SparkConf(false)
    .set("spark.submit.deployMode", config.getDeployMode)
    .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
    .set("spark.memory.offHeap.enabled", (config.getOffHeapMemoryGb > 0).toString )
    .set("spark.memory.offHeap.size", (config.getOffHeapMemoryGb * 1024 * 1024 * 1024).toString)

  @Bean
  def sc(config: SparkProperties, sparkConf: SparkConf): SparkContext = new SparkContext(config.getMaster, config.getAppName, sparkConf)

  @Bean
  def sqlContext(sc: SparkContext): SQLContext = new SQLContext(sc)

  @Bean
  def session(sqlContext: SQLContext): SparkSession = sqlContext.sparkSession

  @Bean
  def scc(sc: SparkContext): StreamingContext = new StreamingContext(sc, Duration(2000))

  @Bean
  @ConditionalOnProperty(Array("kafka.enabled"))
  def stream(config: KafkaProperties, ssc: StreamingContext): InputDStream[ConsumerRecord[String, String]] = {
    var kafkaParams = Map[String, Object](
      ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG -> config.getBootstrapServers,
      ConsumerConfig.GROUP_ID_CONFIG -> config.getConsumerGroup,
      ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG -> classOf[StringDeserializer],
      ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG -> classOf[StringDeserializer]
    )

    KafkaUtils.createDirectStream[String, String](ssc, LocationStrategies.PreferConsistent,
      ConsumerStrategies.Subscribe[String, String](config.getTopics.toSet, kafkaParams))

  }
}